### ISSUE TITLE
pm2 restart overwrites env variables with global env variables

### ISSUE THREAD

**Issue by aronwoost** [Unitech/pm2#1283]
My app json:
```
{
  "name": "app",
  "script": "index.js",
  "instances": 1,
  "exec_mode": "cluster_mode",
  "env": {
    "MY_SETTING": "some_path"
  }
}
```
If global environment variable `MY_SETTING` is set to `"some_other_path"` the app uses the global one (`"some_other_path"`) after `pm2 restart app`.

Interesting fact: when max_memory_restart caused the restart, pm2 uses the "local" env variable (`"some_path"`).

**Comment by jshkurti** [comment]
Can you try it with `npm i -g git://github.com/Unitech/PM2#development` ?
I think this might have been fixed.

**Comment by aronwoost** [comment]
Nope, bug is still there. Can you reproduce?

**Comment by aronwoost** [comment]
Can I help to reproduce/fix the bug?

**Comment by mehwww** [comment]
+1 Same problem hele

**Comment by ThomasBormans** [comment]
+1

**Comment by aronwoost** [comment]
I see that the issue persists in 0.14.3. Problem is, with this issue `pm2 restart pm2.json` break my apps.

Is there anything I can do? Maybe you can point me to somewhere in the codebase so I can have a lock it it? Also I can try to create a test case with some guidenes regarding the pm2 test setup.

Let me know if I can help in any way.

**Comment by soyuka** [comment]
#1366 related?

@aronwoost my test case https://github.com/Unitech/PM2/issues/1366#issuecomment-113515165

Thing is it'd need a lot of refactoring: 
- `--env` is only used by `pm2 start` (see [CLI.js](https://github.com/Unitech/PM2/blob/master/lib/CLI.js))
- `pm2 restart|reload|gracefulReload` would need a refactor to handle `env` properly
- `CLI._reloadProcessName` needs an env parameter (maybe a refactor with `fn(arg, option, cb)` would be better everywhere to simplify further updates)
- This function is called in a lot of places

Feel free to dig in the code, I don't have the time atm :/. 

For the code base, you should know that:
- `bin` is calling `CLI`
- `CLI` is calling `Satan`
- `Satan` is calling `God`

`God` can only be debugged through the pm2 internal logs, the best way to do this is to run in `--no-daemon` mode with `DEBUG="pm2:*"`.

**Comment by aronwoost** [comment]
As a follow up: max-memory-restart does restart the app with the initial ENV. Why is this so?

**Comment by pkyeck** [comment]
:+1: would be nice to have a solution for this

**Comment by Unitech** [comment]
It's normal.

pm2 start process.json -> Will start the app and will overwrite global env with what is declared in env {}
pm2 restart app-name  -> Will take the global environment and will re-assign variables

You need to do:

pm2 restart process.json

But you say that it breaks your apps?

**Comment by pkyeck** [comment]
> pm2 restart app-name -> Will take the global environment and will re-assign variables

when is this the desired action? wouldn't one always (re)start the process as before?

**Comment by Unitech** [comment]
```bash
$ pm2 start app.js
# I want to activate DEBUG flags
$ DEBUG="*" pm2 restart app.js
# Deactivate
$ DEBUG="" pm2 restart app.js
```

**Comment by pkyeck** [comment]
ok, I get that. but shouldn't we then just say that "inline" ENV variables overwrite JSON+global? but keeps the other variables as before? you could then do something like this:

```bash
$ pm2 start conf.json --name=test
# I want to activate DEBUG flags
$ DEBUG="*" pm2 restart test
# Deactivate
$ pm2 restart test
```

**Comment by soyuka** [comment]
+1 @pkyeck it's seems to me that it'd be the logic behavior.

It's something we need to clear out as there are many issues on the subjet, also many different ways of using environnement variables.

**Comment by aronwoost** [comment]
Ok, so why does `"env":{}` does not survive `pm2 restart` when other properties like `"node_args"` and `"log_date_format"` (and others) do? This is totally surprising if not confusing IMHO.

Why this is a problem? Well, a dev/ops can't simply ssh into a machine, `pm2 l` and `pm2 restart [app-index]`. Instead the person has to figure out where the app lives, `cd` into that directory, find the correct json, and then - finally - `pm2 restart app.json`.

Does this make any sense?

**Comment by itnick** [comment]
@aronwoost +1
Same pain and confusion as a devops.
The problem is ```pm2 info``` doesn't give you any hint on which app.json has been used.

**Comment by teddychan** [comment]
+1

Totally agree on @aronwoost said, its super pain esp the dev/ops don't have SSH to the machine.

**Comment by jesus-** [comment]
+1

**Comment by krokofant** [comment]
@aronwoost Described the problem well. If the app environment is setup in the *app.json* and the *pm2 start app.json* starts the app in a certain environment *I assumed* that the restart command would restart the app with the all the previous settings and the same environment.

**Comment by Unitech** [comment]
Everything will be updatable on next release (1.0.3) (environment, configuration attributes...) on restart AND reload.

To avoid updating with the system environment the option `--skip-env` will be available

**Comment by Unitech** [comment]
Update available (PM2@1.1.1 - latest stable):

```
$ npm install pm2@latest -g
$ pm2 update
```

Please re-open this issue if you notice any problem,

**Comment by itnick** [comment]
@Unitech There's no release of `1.0.3`
I tried both pm2 versions `1.0.1` and `1.1.3`

Same issue still exist:
pm2 uses the "local" env variable when I do `pm2 restart`
